## 2. Análisis Exploratorio Inicial (EDA)

Trás definir el objetivo de la práctica procedemos a hacer el analisis exploratorio de los datos. 

En primer luga cargamos todas las librerias que necesitaremos y hacemos una visualización general de dataframe, de su composición y del tipo de datos que contiene cada columna.

In [1]:
import pandas as pd 
import numpy as np
from plotnine import ggplot, aes, geom_line, geom_point, geom_bar, geom_boxplot
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
import missingno as msno
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import KFold
from sklearn.ensemble import RandomForestRegressor
from sklearn.feature_selection import SelectKBest
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RepeatedKFold
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error
from scipy import stats
import statsmodels.api as sm
import statsmodels.formula.api as smf
import pwlf

In [2]:
data= pd.read_csv('/home/inma/Master_Data_Science _Contenido/Fundamentos_de_Analisis _de_Datos/Practica/Datos/Melbourne_housing_FULL.csv')
data.head(20)

Suburb              Address  Rooms Type      Price Method  SellerG  \
0   Abbotsford        68 Studley St      2    h        NaN     SS   Jellis   
1   Abbotsford         85 Turner St      2    h  1480000.0      S   Biggin   
2   Abbotsford      25 Bloomburg St      2    h  1035000.0      S   Biggin   
3   Abbotsford   18/659 Victoria St      3    u        NaN     VB   Rounds   
4   Abbotsford         5 Charles St      3    h  1465000.0     SP   Biggin   
5   Abbotsford     40 Federation La      3    h   850000.0     PI   Biggin   
6   Abbotsford          55a Park St      4    h  1600000.0     VB   Nelson   
7   Abbotsford         16 Maugie St      4    h        NaN     SN   Nelson   
8   Abbotsford         53 Turner St      2    h        NaN      S   Biggin   
9   Abbotsford         99 Turner St      2    h        NaN      S  Collins   
10  Abbotsford       129 Charles St      2    h   941000.0      S   Jellis   
11  Abbotsford         124 Yarra St      3    h  1876000.0      S   Nelson   
12  Abbotsford  121/56 Nicholson St      2    u        NaN     PI   Biggin   
13  Abbotsford        17 Raphael St      4    h        NaN      W   Biggin   
14  Abbotsford        98 Charles St      2    h  1636000.0      S   Nelson   
15  Abbotsford     217 Langridge St      3    h  1000000.0      S   Jellis   
16  Abbotsford      18a Mollison St      2    t   745000.0      S   Jellis   
17  Abbotsford   6/241 Nicholson St      1    u   300000.0      S   Biggin   
18  Abbotsford        10 Valiant St      2    h  1097000.0      S   Biggin   
19  Abbotsford  403/609 Victoria St      2    u   542000.0      S   Dingle   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0   3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1   3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2   4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3   4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4   4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
5   4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
6   4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   
7   6/08/2016       2.5    3067.0  ...       2.0  2.0     400.0         220.0   
8   6/08/2016       2.5    3067.0  ...       1.0  2.0     201.0           NaN   
9   6/08/2016       2.5    3067.0  ...       2.0  1.0     202.0           NaN   
10  7/05/2016       2.5    3067.0  ...       1.0  0.0     181.0           NaN   
11  7/05/2016       2.5    3067.0  ...       2.0  0.0     245.0         210.0   
12  7/11/2016       2.5    3067.0  ...       2.0  1.0    4292.0          82.0   
13  7/11/2016       2.5    3067.0  ...       2.0  0.0     230.0         147.0   
14  8/10/2016       2.5    3067.0  ...       1.0  2.0     256.0         107.0   
15  8/10/2016       2.5    3067.0  ...       NaN  NaN       NaN           NaN   
16  8/10/2016       2.5    3067.0  ...       NaN  NaN       NaN           NaN   
17  8/10/2016       2.5    3067.0  ...       1.0  1.0       0.0           NaN   
18  8/10/2016       2.5    3067.0  ...       1.0  2.0     220.0          75.0   
19  8/10/2016       2.5    3067.0  ...       NaN  NaN       NaN           NaN   

    YearBuilt         CouncilArea Lattitude  Longtitude  \
0         NaN  Yarra City Council  -37.8014    144.9958   
1         NaN  Yarra City Council  -37.7996    144.9984   
2      1900.0  Yarra City Council  -37.8079    144.9934   
3         NaN  Yarra City Council  -37.8114    145.0116   
4      1900.0  Yarra City Council  -37.8093    144.9944   
5         NaN  Yarra City Council  -37.7969    144.9969   
6      2014.0  Yarra City Council  -37.8072    144.9941   
7      2006.0  Yarra City Council  -37.7965    144.9965   
8      1900.0  Yarra City Council  -37.7995    144.9974   
9      1900.0  Yarra City Council  -37.7996    144.9989   
10        NaN  Yar

In [4]:
print("Cantidad de Filas y de Columnas",data.shape)


Cantidad de Filas y de Columnas (34857, 21)
Nombre de las columnas Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')


In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34857 entries, 0 to 34856
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         34857 non-null  object 
 1   Address        34857 non-null  object 
 2   Rooms          34857 non-null  int64  
 3   Type           34857 non-null  object 
 4   Price          27247 non-null  float64
 5   Method         34857 non-null  object 
 6   SellerG        34857 non-null  object 
 7   Date           34857 non-null  object 
 8   Distance       34856 non-null  float64
 9   Postcode       34856 non-null  float64
 10  Bedroom2       26640 non-null  float64
 11  Bathroom       26631 non-null  float64
 12  Car            26129 non-null  float64
 13  Landsize       23047 non-null  float64
 14  BuildingArea   13742 non-null  float64
 15  YearBuilt      15551 non-null  float64
 16  CouncilArea    34854 non-null  object 
 17  Lattitude      26881 non-null  float64
 18  Longti

Durante estas primeras observaciones ya se puede identificar que el dataset  es bastante grande con 34857 y 21 vcolumnas, pero también contiene datos faltantes en bastantes variables incluyendo la varibable objetivo "Price". 

Además, cuenta que  varibles cualitativas y cuantitivas. 
Para poder clasificar los tipo de varibles vamos a realizar una función que lo identifique. 


In [11]:
def esDeProporcion(minimo):
    if(minimo<0):
        return(False)
    else:
        return(True)

def esAcotada(minimo,maximo):
    if( minimo==0 and maximo ==100):
        return(True)
    else:
        return(False)

def clasificar_variables(dataframe):
        
    tipos_columnas=dict()
    for index,i in enumerate(dataframe.keys()):
        if(isinstance(dataframe.iloc[0][index], ( np.int64))):
            #variable cuantitativa discreta
            if(len(dataframe[str(i)].value_counts())>2):
                #politomizada
                minimo=dataframe[str(i)].min()
                maximo=dataframe[str(i)].max()
                if(esDeProporcion(minimo)):
                    if(esAcotada(minimo,maximo)):
                        tipos_columnas[str(i)]='Cuantitativa Discreta de Proporcion Politomizada Acotada'
                    else:
                        tipos_columnas[str(i)]='Cuantitativa Discreta de Proporcion Politomizada No Acotada'
                else:
                    if(esAcotada(dataframe[str(i)])):
                        tipos_columnas[str(i)]='Cuantitativa Discreta de Intervalo Politomizada Acotada'
                    else:
                        tipos_columnas[str(i)]='Cuantitativa Discreta de Intervalo Politomizada No Acotada'

            elif(len(dataframe[str(i)].value_counts())==2):
                #dicotomizada
                tipos_columnas[str(i)]='Cuanitativa Dicotoma'
            else:
                print('deberias borrar la variable '+ str(i))
        elif(isinstance(dataframe.iloc[0][index], ( np.float64))):
            minimo=dataframe[str(i)].min()
            maximo=dataframe[str(i)].max()
            #variable cuantitativa continua
            if(esDeProporcion(minimo)):
                    if(esAcotada(minimo,maximo)):
                        tipos_columnas[str(i)]='Cuantitativa Continua de Proporcion Politomizada Acotada'
                    else:
                        tipos_columnas[str(i)]='Cuantitativa Continua de Proporcion Politomizada No Acotada'
            else:
                if(esAcotada(dataframe[str(i)])):
                        tipos_columnas[str(i)]='Cuantitativa Continua de Intervalo Politomizada Acotada'
                else:
                        tipos_columnas[str(i)]='Cuantitativa Continua de Intervalo Politomizada No Acotada'
        elif(isinstance(dataframe.iloc[0][index],bool)):
            #variable cualitativva dicotoma
            tipos_columnas[str(i)]='Cualitativa Dicotoma'
        elif(isinstance(dataframe.iloc[0][index],str)):
            if(len(dataframe[str(i)].value_counts())>2 and len(dataframe[str(i)].value_counts())<12):
                #politomizada y alomejor ordinal
                print(dataframe[str(i)].value_counts())
                ordinal=input('¿Es ordinal?(seleccionar si o no): ')
                if(ordinal=='si'):
                    tipos_columnas[str(i)]='Cualitativa Politoma Ordinal'
                elif(ordinal=='no'):
                    tipos_columnas[str(i)]='Cualitativa Politoma No ordinal'
                else:
                    tipos_columnas[str(i)]='Cualitativa Politoma Nan'
            elif(len(dataframe[str(i)].value_counts())>2):
                 tipos_columnas[str(i)]='Cualitativa Politoma'
            elif(len(dataframe[str(i)].value_counts())==2):
                #dicotomizada
                tipos_columnas[str(i)]='Cualitativa Dicotoma'
            else:
                print('deberias borrar la variable '+ str(i))
    return(tipos_columnas)

In [24]:
esDeProporcion(-1)
esAcotada(-1,100)
clasificar_variables(data)
   

h    23980
u     7297
t     3580
Name: Type, dtype: int64
¿Es ordinal?(seleccionar si o no): no
S     19744
SP     5095
PI     4850
VB     3108
SN     1317
PN      308
SA      226
W       173
SS       36
Name: Method, dtype: int64
¿Es ordinal?(seleccionar si o no): no


TypeError: esAcotada() missing 1 required positional argument: 'maximo'

En funcion de la clasificación por parte de la función anterior procedemos a analizar las variables cualitativas y cuatitativas por separado. 